plugin/src/main/java/org/opennms/timeseries/impl/influxdb/InfluxdbWriter.java

```diff
@@ -0,0 +1,166 @@
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.ConnectionPool;
+import okhttp3.Dispatcher;
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.RequestBody;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+
+/**
+ * We implement our own write logic with OKClient in order to let our caller handle backpressure.
+ */
+public class InfluxdbWriter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(InfluxdbWriter.class);
+
+    final InfluxdbConfig config;
+    private final OkHttpClient client;
+
+    private final MetricRegistry metrics = new MetricRegistry();
+    private final Meter samplesWritten = metrics.meter("samplesWritten");
+    private final Meter samplesLost = metrics.meter("samplesLost");
+
+    private final Bulkhead asyncHttpCallsBulkhead;
+
+    public InfluxdbWriter(final InfluxdbConfig config) {
+        this.config = config;
+        ConnectionPool connectionPool = new ConnectionPool(config.getMaxConcurrentHttpConnections(), 5, TimeUnit.MINUTES);
+        Dispatcher dispatcher = new Dispatcher();
+        dispatcher.setMaxRequests(config.getMaxConcurrentHttpConnections());
+        dispatcher.setMaxRequestsPerHost(config.getMaxConcurrentHttpConnections());
+
+        this.client = new OkHttpClient.Builder()
+                .readTimeout(config.getReadTimeoutInMs(), TimeUnit.MILLISECONDS)
+                .writeTimeout(config.getWriteTimeoutInMs(), TimeUnit.MILLISECONDS)
+                .dispatcher(dispatcher)
+                .connectionPool(connectionPool)
+                .build();
+
+
+        BulkheadConfig bulkheadConfig = BulkheadConfig.custom()
+                .maxConcurrentCalls(config.getMaxConcurrentHttpConnections() * 6)
+                .maxWaitDuration(Duration.ofMillis(config.getBulkheadMaxWaitDurationInMs()))
+                .fairCallHandlingStrategyEnabled(true)
+                .build();
+        asyncHttpCallsBulkhead = Bulkhead.of("asyncHttpCalls", bulkheadConfig);
+
+        // Expose HTTP client statistics
+        metrics.register("connectionCount", (Gauge<Integer>) () -> client.connectionPool().connectionCount());
+        metrics.register("idleConnectionCount", (Gauge<Integer>) () -> client.connectionPool().idleConnectionCount());
+        metrics.register("queuedCallsCount", (Gauge<Integer>) () -> client.dispatcher().queuedCallsCount());
+        metrics.register("runningCallsCount", (Gauge<Integer>) () -> client.dispatcher().runningCallsCount());
+        metrics.register("availableConcurrentCalls", (Gauge<Integer>) () -> asyncHttpCallsBulkhead.getMetrics().getAvailableConcurrentCalls());
+        metrics.register("maxAllowedConcurrentCalls", (Gauge<Integer>) () -> asyncHttpCallsBulkhead.getMetrics().getMaxAllowedConcurrentCalls());
+    }
+
+    public void writePoints(final List<Point> points) {
+        asyncHttpCallsBulkhead.executeCompletionStage(() -> writeAsync(points)).whenComplete((r, ex) -> {
```

Comment:
Let's prepare the request as much as we can in the calling thread before applying the bulkhead.

Instead of calling `writeAsync(List<Point> points)` change this to `executeAsync` like is done here: https://github.com/OpenNMS/opennms-cortex-tss-plugin/blob/master/plugin/src/main/java/org/opennms/timeseries/cortex/CortexTSS.java#L208

Reply:
fixed